packages/flutter/test/services/hardware_keyboard_init_test.dart

```diff
@@ -0,0 +1,168 @@
+//
+// For simplicity, only the `type` is processed.
+ByteData packRawKeyEvent(KeyData key) {
+  final Map<String, Object?> message = <String, Object?>{
+    'keymap': 'macos',
+    'type': key.type == ui.KeyEventType.up ? 'keyup' : 'keydown',
+    // Leave the rest empty.
+  };
+  return const JSONMessageCodec().encodeMessage(message)!;
+}
+
+// Constructs a `KeyData` with the given type and timestamp.
+KeyData fakeKeyData(ui.KeyEventType type, int timestampInMicroseconds) {
+  return KeyData(
+    timeStamp: Duration(microseconds: timestampInMicroseconds),
+    type: type,
+    physical: PhysicalKeyboardKey.keyA.usbHidUsage,
+    logical: LogicalKeyboardKey.keyA.keyId,
+    character: null,
+    synthesized: false,
+  );
+}
+
+// Send the given key data to both key channels.
+void sendKeyData(KeyData key, {
+  PlatformMessageResponseCallback? keyDataCallback,
+  PlatformMessageResponseCallback? rawKeyEventCallback,
+}) {
+  ui.channelBuffers.push(
+    'flutter/keydata',
+    packKeyData(key),
+    keyDataCallback ?? (_){},
+  );
+  ui.channelBuffers.push(
+    'flutter/keyevent',
+    packRawKeyEvent(key),
+    rawKeyEventCallback ?? (_){},
+  );
+}
+
+void main() {
+  test('Key events before binding is initialized should be ignored', () async {
+    // This test constructs a case where if key events before binding
+    // initialization are not ignored, the event stream will not conform the
+    // event rule (downs and ups must alternate) and thus causes an error.
+
+    FlutterError.onError = (FlutterErrorDetails details) {
+      // Throws if the event rule is broken.
+      expect(true, false);
+    };
+
+    ensureBindingIsNotInitialized();
+
+    // Send event #1, a key down event. This event should be ignored.
+    sendKeyData(
+      fakeKeyData(ui.KeyEventType.down, 1),
+      keyDataCallback: (ByteData? response) { expect(response, null); },
+      rawKeyEventCallback: (ByteData? response) { expect(response, null); },
+    );
+    // Send event #2, a key up event. This event should be ignored, too.
+    //
+    // The reason to send two events before binding is initialized is to prevent
+    // the case where the first event is stored in the message queue (the
+    // default behavior) and therefore the event rule is still conformed even if
+    // pre-initialization events are not ignored (as described in
+    // https://github.com/flutter/flutter/issues/125975 ).
+    sendKeyData(
+      fakeKeyData(ui.KeyEventType.up, 2),
+      keyDataCallback: (ByteData? response) { expect(response, null); },
+      rawKeyEventCallback: (ByteData? response) { expect(response, null); },
+    );
+
+    ensureBindingIsNotInitialized();
+    WidgetsFlutterBinding.ensureInitialized();
+
+    // Wait until the binding starts listening to the event callbacks.
+    //
+    // There isn't an elegant way to listen to this, so this test has to poll.
+    while (true) {
+      await Future<void>.delayed(const Duration(milliseconds: 20));
```

Comment:
we should definitely never write code like this :-)

doesn't the framework synchronously add its listeners? (if not, why not?)